### DISCUSSION TITLE
现各个视频页面（视频、番剧、课程）都能在 `window` 下找到一致的播放器 API

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#4341] [分享 / Share]
- 视频页面：`window.player`, `window.nano`
- 番剧页面：`window.player`, `window.nano`
- 课程页面：`window.playerRaw`, `window.nano`

使用示例：
- `player.seek(60)`：跳转进度到第 60 秒
- `player.goto(1)`：跳转到列表第 2 个视频
- `player.getManifest()`：获取包括 aid、bvid、cid、seasonId、episodeId 等视频信息（see: #4330）
- `player.setHandoff(nano.HandoffKind.Auto)`：开启自动连播
- `player.setHandoff(nano.HandoffKind.Abort)`：关闭自动连播
- `player.on(nano.EventType.Player_Initialized, () => { /* ... */ }) `：监听视频开始播放
- `player.once(nano.EventType.Player_Initialized, () => { /* ... */ }) `：监听视频开始播放，只执行一次
- `player.off(nano.EventType.Player_Initialized, listener) `：停止监听

**Comment by WakelessSloth56** [comment]
- `async player.requestQuality(qulity:number)`：切换视频画质，参数参见[清晰度标识](https://github.com/SocialSisterYi/bilibili-API-collect/blob/master/docs/video/videostream_url.md#qn%E8%A7%86%E9%A2%91%E6%B8%85%E6%99%B0%E5%BA%A6%E6%A0%87%E8%AF%86)。
  - 同步更新 UI （会弹出“正在切换”、“已切换至”等 Toast，播放器控制栏中的选择也会更新）

- `player.getSupportedQualityList()`：获取当前视频所有的清晰度，例如：`[120, 116, 80, 64, 32, 16]`。

- `player.getQulity()`：获取画质、之前的画质等，例如：
  ```javascript
    {
      "oldA": null,
      "nowA": 0,
      "newA": 0,
      "oldQ": 112,
      "nowQ": 80,
      "newQ": 80,
      "realQ": 80
    }
  ```

**Reply by feeas** [reply]
函数打错了,是getQuality

**Comment by WakelessSloth56** [comment]
- `player.setPreferCodec(codec:number)`：设置视频编码策略。
- `player.getPreferCodec()`：获取当前视频编码策略。

此数据存储在 本地存储(localStorage)->`https://www.bilibili.com`->`bilibili_player_codec_prefer_type` 中。

|代码|视频编码|
|-|-|
|`0`|默认|
|`1`|`HEVC/H.265`|
|`2`|`AVC/H.264`|
|`3`|`AV1`|

注：使用此接口切换不会同步更新 UI 的显示，UI 中高亮选项是通过 CSS 的 `input:checked` 实现。

**Comment by WakelessSloth56** [comment]
**此方法已失效（#4795）**

- 获取播放器全部字幕设置： `player.__core().subtitleStore.state`

<details>

示例：

```json5
{
    "hover": false,
    "enable": true,
    "isclosed": false,
    "lan": "zh-CN", // 主字幕语言
    "minorLan": "ai-zh", // 副字幕语言
    "fade": false,
    "color": "16777215",
    "scale": true,
    "shadow": "0",
    "opacity": 0.4,
    "fontsize": 1,
    "bilingual": true, // 双语字幕
    "position": "bottom-center",
    "languageList": [
        {
            "id": 0, // ...
            "lan": "zh-CN",
            "lan_doc": "中文（中国）",
            "is_lock": false,
            "subtitle_url": "...", // 字幕地址
            "type": 0,
            "id_str": "...", // 注：和 id 不一样
            "ai_type": 0,
            "ai_status": 0
        },
        {
            "id": 0, // ...
            "lan": "ai-zh",
            "lan_doc": "中文（自动生成）",
            "is_lock": false,
            "subtitle_url": "...",
            "type": 1,
            "id_str": "...",
            "ai_type": 0,
            "ai_status": 2
        }
    ]
}
```

</details>
